Architecture.md

```diff
@@ -0,0 +1,165 @@
+<!-- This file contains a rough overview of the PyO3 codebase. -->
+<!-- Please do not make descriptions too specific, so that we can easily -->
+<!-- keep this file in sync with the codebase. -->
+
+# PyO3: Architecture.md
+
+This document roughly describes the high-level architecture of PyO3.
+If you want to become familiar with the codebase, you are in the right place!
+
+## Overview
+
+PyO3 provides a bridge between Rust and Python, based on the [Python C/API].
+Thus, PyO3 has low-level bindings of these API as its core.
+On top of that, we have higher-level bindings to operate Python objects safely.
+Also, to define Python classes and functions in Rust code, we have `trait PyClass<T>` and a set of
+protocol traits (e.g., `PyIterProtocol`) for supporting object protocols (i.e., `__dunder__` methods).
```

Comment:
```suggestion
protocol traits (e.g. `PyIterProtocol`) for supporting object protocols (i.e. `__dunder__` methods).
```

Reply:
Maybe American or British choices? https://english.stackexchange.com/a/6896

Reply:
Ah interesting, I wasn't aware that the Americans did it differently to what I'm used to. In which case either seems fine to me, let's keep what you've written!